### ISSUE TITLE
Issue with web-launch-url Parameter Handling on Windows in VS Code

### ISSUE THREAD

**Issue by AoEiuV020** [Dart-Code/Dart-Code#5322]
**Describe the bug**
When using web-launch-url in VS Code on Windows, it fails to pass multiple parameters correctly to the Flutter web app. Despite identical configuration files, this only occurs on Windows, while Linux and macOS handle the parameters as expected.
Details
Using the following configuration from my launch.json file:
```
         {
            "name": "example(livekit)",
            "cwd": "example",
            "request": "launch",
            "type": "dart",
            "env": {
                "DISPLAY": ":0"
            },
            "args": [
                "--web-browser-flag=--disable-web-security",
                "--web-port=9080",
                "--web-launch-url=http://localhost:9080/?autoConnect&serverUrl=https%3A%2F%2Fmeet.livekit.io&room=123456&name=vscode",
                "--dart-entrypoint-args=--autoConnect",
                "--dart-entrypoint-args=--serverUrl=https://meet.livekit.io",
                "--dart-entrypoint-args=--room=123456",
                "--dart-entrypoint-args=--name=vscode",
            ]
        }
```
https://github.com/AoEiuV020/livekit_meeting/blob/89b0919e4792cfbde3d7f8407733406d30eea1ff/.vscode/launch.json#L30
On Windows, the browser URL launched in the address bar is truncated to:

```
http://localhost:9080/?autoConnect
```
All parameters after & are lost, and when targeting Windows, this configuration also causes all subsequent dart-entrypoint-args to be ignored. However, if the web-launch-url is removed, the dart-entrypoint-args function normally.

Running flutter run directly from the command line with parameters works without issue:
```
flutter run -d chrome "--web-browser-flag=--disable-web-security" "--web-port=9080" "--web-launch-url=http://localhost:9080/?```
autoConnect&serverUrl=https%3A%2F%2Fmeet.livekit.io&room=123456&name=vscode"
```

**To Reproduce**

1. In VS Code, select the Chrome device as the target.
2. Go to the Run and Debug view.
3. Select the example (livekit) configuration.
4. Click on the Debug button to start the Flutter web app.

**Expected behavior**
The web-launch-url should consistently retain all query parameters across platforms when launching a Flutter web app from VS Code.

**Screenshots**
![image](https://github.com/user-attachments/assets/9295f802-d204-42fe-afb3-7f8430b5d8a4)

**Please complete the following information:**

[log-1f1f.md](https://github.com/user-attachments/files/17538396/log-1f1f.md)

 - Operating System and version: win11
 - VS Code version: 1.94.2
 - Dart extension version: 3.98.1
 - Dart/Flutter SDK version: 3.24.4
 - Target device (if the issue relates to Flutter debugging): chrome 129.0.6668.101

**Comment by DanTup** [comment]
I can reproduce this, but I also see it from a PowerShell terminal:

```
flutter run "-dchrome" "--web-port=9080" "--web-launch-url=http://localhost:9080/?a&b=c"
```

https://github.com/user-attachments/assets/a3d0b55d-0d17-400c-aa23-775e9aa95985


However, if I add `--%` to prevent PowerShell parsing the args, it's good:

```
flutter --% run "-dchrome" "--web-port=9080" "--web-launch-url=http://localhost:9080/?a&b=c"
```

We're just using `Process.start()` for this, so I'll need to figure out what escaping Dart is (or isn't) doing for shell execution (since Flutter is a `.bat`, it has to execute via a shell).

**Comment by DanTup** [comment]
Looks like this is documented:

https://api.dart.dev/stable/3.5.4/dart-io/Process/start.html

> **NOTE**: On Windows, if `executable` is a batch file (`'.bat'` or `'.cmd'`), it may be launched by the operating system in a system shell regardless of the value of `runInShell`. This could result in arguments being parsed according to shell rules. For example:
>
>```
> void main() async {
>   // Will launch notepad.
>   Process.start('test.bat', ['test&notepad.exe']);
> }
> ```

Looks like the VM has a private implementation of argument escaping here that perhaps we can use:

https://github.com/dart-lang/sdk/blob/6d4ebea83b7217b66112df7268ffc5fa6dd2fd36/sdk/lib/_internal/vm/bin/process_patch.dart#L342

**Comment by DanTup** [comment]
I'm actually struggling to track down what's happening here. I wrote a small test script that creates ` .bat` file that calls a Dart script and passes the args with `%*` (which is what Flutter does), then calls it with a load of strings with various characters in. Without any escaping, they all come through correctly. With escaping, the escaped values come through to Dart:

```dart
import 'dart:convert';
import 'dart:io';

const testCharacters = r'"&:\<>|%=+-_@~#<>?/';

final testArgs = [
  for (var char in testCharacters.split('')) 'one${char}two',
];

Future<void> main(List<String> args) async {
  if (args.isNotEmpty && args.first == 'printargs') {
    printArgs(args.skip(1).toList());
    return;
  }

  // Create a .bat file that just forwards all args back to us.
  File('tool/print_args.bat').writeAsStringSync('dart tool/test_args.dart %*');

  var proc = await Process.start(
    Platform.resolvedExecutable,
    [
      '--enable-asserts',
      Platform.script.toFilePath(),
      'printargs',
      ...testArgs,
    ],
  );

  proc.stdout.transform(utf8.decoder).listen(print);
  proc.stderr.transform(utf8.decoder).listen(print);

  print('Started process ${proc.pid}.. waiting for exit');
  await proc.exitCode;
}

void printArgs(List<String> args) {
  assert(args.length == testArgs.length);
  for (var i = 0; i < args.length; i++) {
    var a = testArgs[i];
    var b = args[i];
    print('$a came through as $b    ${a == b ? '' : 'MISMATCH'}');
  }
}
```

![image](https://github.com/user-attachments/assets/d2e3d33b-96ca-424d-b50a-80b350a608d0)

It's not clear to me why this works given the docs quoted above (and the actual bug reported here). It's going to need some more digging.

**Comment by DanTup** [comment]
I managed to repro this with a Dart script when the executable path contains a space (my Flutter SDK is inside a folder named "Flutter (main)" to catch some issues with escaping like this!). This seems to match some of the description here from the legacy DAP in TS (second line):

https://github.com/Dart-Code/Dart-Code/blob/aac661885cffcd1695af51ba6d0f01259f8a1b7a/src/shared/processes.ts#L30-L40

However, moving my Flutter SDK to a folder without a space doesn't prevent the original issue from happening, which seems slightly odd.

In any case, the correct fix here is probably for DAP to escape all of the args and quote them. It's unfortunate because this escaping is probably shell-specific and the rules don't seem complex.

(@bkonyi FYI... it would sure be nice if Flutter didn't require a shell script to invoke 😄)

**Comment by bkonyi** [comment]
Yeah, I _really_ don't like that `flutter.bat` and `dart.bat` (and their `.sh` equivalents) exist. It'd probably be difficult to migrate away from them and still support shipping Flutter via GitHub, so I doubt it'll change any time soon. It's _probably_ safe to assume those artifacts exist if the plugin already knows it's run something like `flutter doctor` to initialize the SDK, but poking directly at snapshots is something we discourage since we do change their names from time to time which causes headaches when tooling is reaching into SDK internals.

**Comment by DanTup** [comment]
Right, I think I have a good isolated repro for this now...

The script below fails with:

```
Writing scripts to Directory: 'C:\Users\danny\AppData\Local\Temp\flutter dap args escape test1ea07fdc'
Executing:
  executable: C:\Users\danny\AppData\Local\Temp\flutter dap args escape test1ea07fdc\fake_flutter.bat
  args: beforeaafter beforebafter beforecafter before&after
  runInShell: true
'after' is not recognized as an internal or external command,
operable program or batch file.
Expected: ['beforeaafter', 'beforebafter', 'beforecafter', 'before&after']
  Actual: ['beforeaafter', 'beforebafter', 'beforecafter', 'before']
   Which: at location [3] is 'before' instead of 'before&after'
package:matcher             expect
test\escape_test.dart 65:5  main.<fn>
escape_test.dart:65
✖ test escaping

Exited (1).
```

This seems to match what we're seeing with `&` in Flutter. I found some rules at https://stackoverflow.com/questions/6427732/how-can-i-escape-an-arbitrary-string-for-use-as-a-command-line-argument-in-windo but in my testing it doesn't seem to work correctly (for ex. doubling up backslashes seems to not work).

I'll do some more testing and see if I can come up with some escaping rules that pass for all of those characters and then test it in Flutter.


```dart
import 'dart:convert';
import 'dart:io';
import 'package:path/path.dart' as path;
import 'package:test/test.dart';

const testCharacters = r'abc&'; // r' "&:\<>|%=+-_@~#<>?/';

final testArgs = [
  for (var char in testCharacters.split('')) 'before${char}after',
];

void main() {
  test('test escaping', () async {
    var tempDir =
        Directory.systemTemp.createTempSync('flutter dap args escape test');
    print('Writing scripts to $tempDir');

    // Write a shell script to simulate the Flutter .bat file and a Dart script
    // that just prints the arguments sent to it one-per-line.
    var tempShellScript = File(path.join(tempDir.path, 'fake_flutter.bat'));
    var tempDartScript = File(path.join(tempDir.path, 'print_args.dart'));
    var command = [
      '"${Platform.resolvedExecutable}"',
      '"${tempDartScript.path}"',
      "%*",
    ];
    tempShellScript.writeAsStringSync('@echo off\n${command.join(" ")}');
    tempDartScript.writeAsStringSync(r'''
void main(List<String> args) {
  print(args.join('\n'));
}
''');

    var executable = tempShellScript.path;
    var args = testArgs;

    print('''
Executing:
  executable: $executable
  args: ${args.join(' ')}
  runInShell: true
''');
    var proc = await Process.start(
      '"$executable"',
      args,
    );

    var stdoutFuture = proc.stdout.transform(utf8.decoder).toList();
    var stderrFuture = proc.stderr.transform(utf8.decoder).toList();
    await proc.exitCode;

    var stdout = (await stdoutFuture).join().trim();
    var stderr = (await stderrFuture).join().trim();

    if (stderr.isNotEmpty) {
      print(stderr);
    }

    var actual = stdout
        .split('\n')
        .map((l) => l.trim())
        .where((l) => l.isNotEmpty)
        .toList();

    expect(actual, testArgs);
  });
}
```

**Comment by DanTup** [comment]
@bkonyi

> but poking directly at snapshots is something we discourage since we do change their names from time to time which causes headaches when tooling is reaching into SDK internals.

Yeah, that's my fear. I kinda wish we had a `dart.exe flutter` command that just escaped the args correctly and forwarded them on to `flutter.bat` so that every editor wanting to call Flutter didn't have to accurately implement all of the weird Windows shell escaping rules themselves (or more likely, have bugs like this one) 😄

Or similar, when Flutter downloads the Dart SDK, it compiles a tiny `flutter.exe` that does this forwarding. They're all hacks, but a centralised hack in the SDK might be better than hacks (or bugs) in lots of other places.

It's annoying because we're _only_ using shell execute because Flutter is a shell script. We don't actually want _any_ difference in how args are treated, it just happens because the command gets squished to a string to be executed.

Another option could be that when `flutter.bat` downloads the SDK, it also compiles a simple `flutter.exe` alongside `flutter.bat` that escapes and forwards

**Comment by DanTup** [comment]
I've not yet found a way to escape `&` correctly here, so I've filed https://github.com/dart-lang/sdk/issues/59604 to see if someone can advise how this should work (adding `^` which I believe should work, does not seem to).

**Comment by Wdestroier** [comment]
Hi @DanTup, I'm not sure if the issue is related, but I'm trying to launch Flutter on http://localhost:3000 and opening Chrome on https://localhost:5173/f/ by setting the `--web-launch-url` flag, but it's launching on http://localhost:3000/f. That's the behavior when I click the  "Run" button above the main method to launch the app.

I want to open Chrome on another URL because I setup a reverse proxy to deal with CORS and Cookies. The following message is printed when I press CTRL+F5 to launch the app:
```
Launching lib\main.dart on Chrome in debug mode...
"https://localhost:5173/f/" is not a valid HTTP URL.
```

Another issue is clicking "Run" or pressing CTRL+F5 leads to different results. For example: I prefer to click the "Run" button, but then Flutter is launched on a random port (completely ignores `launch.json`). I applied this workaround to improve the experience: https://stackoverflow.com/a/76259368.

**Comment by DanTup** [comment]
> Hi @DanTup, I'm not sure if the issue is related, but I'm trying to launch Flutter on http://localhost:3000/ and opening Chrome on https://localhost:5173/f/ by setting the --web-launch-url flag, but it's launching on http://localhost:3000/f. That's the behavior when I click the "Run" button above the main method to launch the app.

No, I don't think it's related - but please file a new issue about this so we can track down where it's being changed.


> Another issue is clicking "Run" or pressing CTRL+F5 leads to different results. For example: I prefer to click the "Run" button, but then Flutter is launched on a random port (completely ignores launch.json)

This is somewhat expected, because pressing F5 will launch your _active_ debug configuration (which can be selected on the Debug sidebar) whereas the Run/Debug links always run a default launch configuration with the appropriate file (and possibly args, if it's a test/group) set. You can override this behaviour by marking a launch configuration (in launch.json) as a `templateFor` - see https://dartcode.org/docs/launch-configuration/.

**Comment by DanTup** [comment]
> "https://localhost:5173/f/" is not a valid HTTP URL.

I misread your original issue thinking the trailing slash was the issue, but now I see it's HTTPS. I believe this is a result of this regex here:

https://github.com/flutter/flutter/blob/099e6d39fe38dbda78dd5efb7d8424bf7fd845a2/packages/flutter_tools/lib/src/web/web_device.dart#L119

I suspect that regex could have `s?` added to it to fix this if it is _only_ used for launching the browser, but if you're also expecting the server to listen on HTTPS (eg. you're not proxying it yourself) then probably it would require work for the internal web server to use HTTPS. This was noted at https://github.com/flutter/flutter/pull/95002#discussion_r768102005.

**Comment by Wdestroier** [comment]
> You can override this behaviour by marking a launch configuration (in launch.json) as a templateFor in which case it can be used for the Run/Debug CodeLens links - see https://dartcode.org/docs/launch-configuration/.

Amazing, it works! It's a great QoL improvement, thank you.

> I suspect that regex could have s? added to it to fix this if it is only used for launching the browser, but if you're also expecting the server to listen on HTTPS (eg. you're not proxying it yourself) then probably it would require work for the internal web server to use HTTPS. 

I was searching the codebase and found this code too. The reverse proxy is adding TLS, not throwing the error would be enough.

Thank you for the answers!
